### DISCUSSION TITLE
[일지] 6월 3일 (목) 오후 회의록 + 멘토링 내용

### DISCUSSION THREAD

**Discussion by olenmg** [bcaitech1/p4-opt-5-vibrhanium-#40] [일지]
###  Q1. data imbalance 문제에 대해서
- A. 기법은 여러가지. 잘 되는 기법은 별로 없음. 결국 쓰이는건 Oversampling, Undersampling. 100:1의 imbalance가 있다면 1 -> 100 혹은 100 -> 1. 정확히 1:1로 맞추지 않고, 그렇다고 100:1은 너무하고. 그 중간을 찾는 것은 heuristic 영역.
Loss를 잘 조작해서 1에 큰 weight을 주는 방법.
augmentation 등 방법을 통해 데이터에 변형을 가해서 데이터 수를 불리는 것이 best. 그러나 도메인에 따라서 (NLP와 같은 경우) augmentation이 힘든 경우에는 데이터를 그냥 복사하기도 함.  
- Q1-1. Weighted loss와 단순 데이터 복사의 차이점?   
  - A. 개념적으로는 똑같고 구현 방식의 차이. 학습 시간의 차원에서 loss를 많이 주는 편이 학습 시간은 빠를 것. 구현상으로 쉬운건 oversampling.  
- Q1-2. 복사한 데이터에 따른 overfitting. 성능 하락은?  
  - A. 실제 모델을 apply할 때의 real 환경에서의 데이터 분포를 보고 판단. 100:1에서 1이나 100이나 중요도가 같은 경우, imbalance를 해결할 필요가 크게 없음. 반면 1이 매우 중요한 경우, imbalance 문제를 해결하는 것이 중요.  
- Q1-3. 경험상 Loss의 변경 (-> Focal loss)이 큰 성능 향상을 보이지 않았음.  
  - A. 데이터 특성에 귀인..? 문제의 특성에 관련있지 않을까.  
- Q1-4. label smoothing, focal loss   
  - A. 그렇게 해석할 수 있어요. 
- Q1-5. nlp에서 imbalance가 감이 안온다. nlp에서는 어떤 형태로 나오나요  
  - A. 보통 class imbalance를 말한다. 감정 분류 문제에서 review를 분석했는데 긍정보다 부정이 훨씬 많은 경우 🤬   
- Q1-6. NLP augmentation을 위해서 한 문장에서 단어를 유의어로 치환하는 방법(HiDEx)를 현업에서 사용하나요?  
  - A. 비슷한 방식으로 Adversarial training을 시도. 딥러닝 모델만 헷갈리는 정보를 삽입해 모델에 혼란을 주어 robust하게 만드는 방법. 이미지 도메인에선 효과가 검증. 그러나 NLP에 적용은 실패. 이미지는 픽셀 값이 살짝 바뀐다고 해서 내용이 크게 바뀌지 않지만 NLP의 경우 단어의 임베딩(컴퓨터) 자체가 인간의 이해와 완벽하게 일치하지 않기에 실패한 듯. HiDEx도 그런 의미에서 잘 되지 않을 것 같다.  
- Q1-7. MRC에도 데이터 임밸런스가 존재하는가?
  - A. MRC같이 클래스 임밸런스가 존재하지 않는 task에서는 보통 데이터 임밸런스라고 하지 않고 데이터 분포가 다른 경우가 존재.

### Q2. KLUE 데이터셋에 대해 어떻게 생각하시는지 원래 데이터셋이 불완전 한건지 보통의 데이터셋이 그런건지 
  - A. 에러가 좀 있지만 KLUE라는 한국어 벤치마킹 데이터셋이 최초로 나왔다는 점에 의의가 있다.

### Q3. 네이버 채용 단계 때 준비는 어떤 방식으로 하셨나요?
  - A. **커뮤니티 활동 (ex. 모두의 연구소)** 인맥 형성(열심히 하면 인상이 좋으니까), 코딩 테스트 준비 (백준, 알고리즘 책), 면접 준비 (통계 공부)
  - Q3-1. 통계공부 어떻게?
    - A. (네이버 GPT 담당) 아래 책을 공부해보세요
      [캐빈머피 - 머신러닝](https://probml.github.io/pml-book/book1.html)
      pattern recognition and machine learning
      PRML은 좀 더 CS 쪽, 머피 머신러닝은 좀 더 통계쪽

### Q4. pretrained된 가중치를 사용할 때는 모델의 모양이 맞아야 가져올 수 있는 것으로 알고 있습니다. 그 가중치를 가져온 모델 구조에서 일부 블록을 빼내어 파라미터를 줄이면서 기존 pretrained된 가중치의 일부를 사용할 수 있는지 궁금합니다!
  - A.
    - 앞의 3개 layer만 사용 -> 작동 잘 함
    - 주의 : 뒷 쪽 weight는 앞 쪽 weight를 바탕으로 만들어진 것. 그래서 뒷 쪽, 중간 쪽 weight만 사용하는 것은 아예 random이 되어버린다.
    - 그거를 넘어서서 모델의 모양이 바뀌는 데 pretrain weight를 쓴다 -> 딱히 생각나는 그림 없음
    - distillation이면 가능할 것으로 생각됨

  - Q4-1. 층마다 포작하는 feature가 다른데, 중간 층은 사용할 수 없을 까요? 뒤에 layer를 가져다가 앞의 layer를 fine tunining
    - 가능할 것 같기도 한데, working을 잘 할 수 있을지 모르겠다.
    
### Q5. 프로젝트 피드백(재영)
  - A.
    - 데이터가 많은 것을 가정하고(computing cost가 많이 들 것이라고 가정하고) 아키텍쳐를 짜는게 좋을 것
    - 학습 데이터로 쓸 수 있는 데이터의 범위?영역?을 잘 설정하기
    - 하고자하는 task(label의 기준)를 명확히 정의
    - 람다는 시간, 용량, ? 제한이 있다.
    - 모델이 작다면 pretrained model을 쓰지 않고 scratch부터 training하는 것도 좋은 방법
    - counter base(단어수 세기)

### Q6. 로봇 프로젝트
  - A. 라즈베리 파이 구매를 추천
    - MB 단위로 모델 줄이는 경우 있음
    ex. yolo image recognition은 라즈베리 파이에 올려서 쓰는 경우 있음
    몇 MB ~ 몇 십 MB

### Q6-1. 핸드폰에 올리는 것은 경량화에 있어서 유의미한 시도일까요? (핸드폰 용량이 많아서 의미 없을까요?)
  - A. 핸드폰에서 돌리는 것은 상당한 기술력이 필요
    - edge device에서 돌리기 위한 기술과 library 경험 필요

### Q6-2. 관련된 한계를 확인하고 요구 조건을 맞추는 방식으로 프로젝트를 진행하느 것 어떻게 생각하시나요?
2주내에는 못하지만, 이정도 하려면 모델 크기가 어느 정도까지 줄어들어야하는 것은 어떨까요?
  - A. 경량화는 vision에서 상당히 앞서나가고 있다. 연구, library 측면 모두 (cf. NLP는 걸음마 단계)
    - 우리나라 vision model을 올리는 기술력 높음
    - 아자르? 최적화 기술 많이 들어간다. **하이퍼 커넥트**. 네이버 데뷰 (Deview..??)에서 발표함. 핸드폰 랜덤 영상 채팅 어플. 실시간으로 영상 필터를 씌워야 해서 최적화가 발달.
    - **스냅챗?**

### Q7. nota 실제로 어떤 일 하나요? Upstage도 어떤 일 하는지 잘 모르겠어요.
  - A.
    - nota는 아는 사람이 없어서 답변이 어렵습니다.
    - upstage는 upstage 분들께 물어보는 게 빠르지 않을까요? 
        - upstage 내에서도 여러 개의 프로젝트가 존재
        - 컨설팅, 교육 프로젝트, 대기업과 협약 맺어서 모델 만들기(B2B?)
        - upstage researcher 분들은 명확한 전문성, 사람들은 :+1:
        - 전문성을 배우고 싶다면 upstage 추천

### Q8. 쓸만한 클라우드 서버
  - A.
    - heroku 
    - AWS s3 가장 싸다(파일 올리고 가져다 쓰는 것 까지는 싸다.)
        - dynamic한 부분 없다면

### 다음 멘토링 - 목요일 8시

---

### 일정
- (종헌) tensor decomposition은 내일 아침까지 구현을 해보겠습니다
- 내일 아름 MobileNet v2 발표 (중간 발표라도) 

### Tensor decomposition
- weight 확인하기 (직접 돌려보기) & MLP로 하기
- AutoML로 일단 랭크 안찾는 방향으로 구현하고 -> 그 다음에 잘 되면 랭크까지 automl로 찾기
    - AutoML에서 모델/하이퍼파라미터 찾기 + rank 탐색까지 한 번에 하는건 힘들 것 같음.
    - 그래서 두 단계를 분해하고, AutoML로 찾은 모델 중 TOP-k개 뽑아서 그것들로 rank 찾고 학습시키는 과정 다시 수행

### 오늘 할 일
- (종헌) DB 연동 코드 추가
- (아름,재경,건우) fire 모듈 추가
- 오늘 밤에 다같이 같은 DB로 돌리기

- config/key는 input에 놓고 각자 설정해서 돌릴 수 있도록/ git이 tracking하지 않도록
